_posts/2017-03-15-Performance-of-LSTM-Using-CuDNN-v5.md

```diff
@@ -0,0 +1,135 @@
+cuDNN provides highly tuned implementations for standard routines such as LSTM, CNN.
+
+In this article, we compare the performance of LSTM with or without cuDNN.
+An LSTM (NStepLSTM) implementation is included in Chainer, which can be found here.
+
+[https://github.com/pfnet/chainer/blob/master/chainer/functions/connection/n_step_lstm.py](https://github.com/pfnet/chainer/blob/master/chainer/functions/connection/n_step_lstm.py)
+
+NVIDIA’s official blog is related to this post. Please check [here](https://devblogs.nvidia.com/parallelforall/optimizing-recurrent-neural-networks-cudnn-5/).
+
+
+
+### Summary: 
+Thorough our experiments, we found the following observations:
+We should use cuDNN
+* if the model is large.
+* if the sequence data is long.
+* if the sequence data has variable lengths. 
+
+
+We conducted experiments from the following viewpoints:
+* The effect of mini-batch size
+* The effect of the number of LSTM layers and sequence length of data
+* The effect of the number of LSTM layers and random sequence length of data
+* The effect of the number of LSTM layers and the input unit size
+* The effect of the number of LSTM layers and the hidden unit size
+* The effect of the number of LSTM layers and dropout rate
+* When will the differences be large? (setting with/without cuDNN)
+
+
+
+### Experimental Result:
+
+* The effect of mini-batch size
+parameters: batchsize = {127, 128, 200, 255, 256}
+In all results, batchsize 128 is faster than 127 and batchsize 256 is faster than 255. (Despite smaller batch size!)
+Using batchsize = 2^n will provide us faster result.
+Note that the number of iterations is the same number (39 iterations) in the case of batchsize = {255, 256}. 
+
+Comparing the setting with/without cuDNN, about 2 times ~ 2.8 times faster when using cuDNN in forward time, and 1.6 times faster in backward time.
+
+ 
+![The effect of mini-batch size]({{ site.url }}/assets/nsteplstm-cudnn-v5-minibatch.png)
+    
+
+* The effect of the layer size of LSTM and sequence length of data
+parameters: length={5, 25, 50}, layer={1, 2, 3}
+In the setting without cuDNN, as the length of data and the number of layers increases, it takes longer time than the setting with cuDNN.
+
+When we use a large LSTM setting (layer=3, length=50), about 7 times faster in forward time, and 4 times faster in backward time with cuDNN.
+
+When we use a small LSTM setting (layer=1, length=5), about 1.5 times faster in forward time, and 1.04 times faster in backward time.
+
+
+![The effect of the layer size of LSTM and sequence length of data]({{ site.url }}/assets/nsteplstm-cudnn-v5-length-layers.png)
+
+* The effect of the layer size of LSTM and random sequence length of data
+parameters: layer = {1, 2, 3}, random={True, False}
+In this setting, we compare if the data sequence length is fixed or not. (Max length is 25.)
+When we use cuDNN, the effect of randomness is small. 
```

Comment:
Consider rewording:
When we use cuDNN, the performance impact of random sequence length is small.